### ISSUE TITLE
Loss to nan

### ISSUE THREAD

**Issue by lizhipro** [sdv-dev/CTGAN#73]
* CTGAN version: 0.2.2
* Python version: 3.6
* Operating System: Linux ubuntu 16.04
* Pytorch 1.1.0, torchversion 0.3.0
* cuda 9.0

### Description

ctgan = CTGANSynthesizer()
ctgan.fit(data, discrete_columns=discrete_columns, epochs=2500)

Epoch 1, Loss G: 1.2144, Loss D: -1.0157
Epoch 2, Loss G: 0.6393, Loss D: -0.0552
Epoch 3, Loss G: 0.3184, Loss D: 0.2490
Epoch 4, Loss G: nan, Loss D: nan
Epoch 5, Loss G: nan, Loss D: nan
Epoch 6, Loss G: nan, Loss D: nan

It is very strange that I use the same code and parameters to run on the CPU and the above situation does not occur.

**Comment by csala** [comment]
Thanks for reporting this @lizhipro I see that this was asked before in #47 , but I'm afraid I have never been able to reproduce it.

There @seansaito mentioned using the census dataset. Is this also the case here? Are you using the census dataset or your own data?

**Comment by lizhipro** [comment]
In fact, i have try in several dataset census, adult, and covtype......, it always appears

---------------
I guess it might be because the version of torchvision is too old, because I don’t have permission to update the nvidia driver of the server, so I have to install the old version of torch and torchvision

**Comment by csala** [comment]
Good point! That could be the issue, indeed, and I did not think about trying it.

I'll try to downgrade torch versions and maybe get my hands on an older CUDA environment to see if I can also reproduce it.

**Comment by csala** [comment]
I can confirm that the torch (or the CUDA) version seems to be the issue. I can reproduce it with these versions:

- torch 1.1.0
- torchvision 0.3.0
- python 3.6
- CUDA (as seen by torch): 9.0.176

**Comment by csala** [comment]
Using torch 1.2.0 or higher, which already use CUDA 10, it seems to be working fine.

**Comment by lizhipro** [comment]
Thank you very very much for taking the time to reproduce the bug. I don’t know if you have a way to avoid this problem by modifying a small amount of code, because I currently have no permission to update the nvidia driver, so I can’t update cuda

**Comment by csala** [comment]
The culprit seems to be this issue: https://github.com/pytorch/pytorch/issues/22442

The problem shows up here: https://github.com/sdv-dev/CTGAN/blob/68a0661f2ef31e2b0f019cabed791457ae5e8b08/ctgan/synthesizer.py#L58

Sometimes this `gumbel_softmax` call returns `nan` values and this screws the optimization for the rest of the training process.

If you want to work around it temporarily you can easily add a health check around that line, and if the `gumbel_softmax` returns a `nan` simply retry the call several times until it produces clean values:

```python
transformed = functional.gumbel_softmax(data[:, st:ed], tau=0.2)
while torch.isnan(transformed).any():
    # You may want to add some logic here to prevent an infinite loop
    # and raise an error if NaNs keep being returned in all the iterations
    transformed = functional.gumbel_softmax(data[:, st:ed], tau=0.2)

data_t.append(transformed)
```

I don't think we will be working on this on our side, since this is an old issue that has already been resolved, but if you implement it and want to make a PR it will be welcome!

**Comment by lizhipro** [comment]
Got it, Thank you so much !

**Comment by seansaito** [comment]
Hi @csala and @lizhipro , thank you for looking into this issue!

As mentioned above, it appears that `functional.gumbel_softmax` is unstable for older versions of pytorch (I also had a similar restriction of not being able to upgrade!). I had implemented a workaround similar to the one above in my fork which fixed the issue.

https://github.com/seansaito/CTGAN/blob/0855c4aef9b458a1cc4f24b7be480fd7a2d346b2/ctgan/synthesizer.py#L74
